game-core/src/main/java/games/strategy/triplea/ai/tree/BattleStep.java

```diff
@@ -0,0 +1,1216 @@
+
+  // Taken from MustFightBattle::returnFireAgainstAttackingSubs
+  private MustFightBattle.ReturnFire returnFireAgainstAttackingSubs(
+      final Collection<Unit> attackingUnits, final Collection<Unit> defendingUnits) {
+    final boolean attackingSubsSneakAttack =
+        defendingUnits.stream().noneMatch(Matches.unitIsDestroyer());
+    final boolean defendingSubsSneakAttack =
+        defendingSubsSneakAttackAndNoAttackingDestroyers(attackingUnits);
+    final MustFightBattle.ReturnFire returnFireAgainstAttackingSubs;
+    if (!attackingSubsSneakAttack) {
+      returnFireAgainstAttackingSubs = MustFightBattle.ReturnFire.ALL;
+    } else if (defendingSubsSneakAttack || Properties.getWW2V2(parameters.getData())) {
+      returnFireAgainstAttackingSubs = MustFightBattle.ReturnFire.SUBS;
+    } else {
+      returnFireAgainstAttackingSubs = MustFightBattle.ReturnFire.NONE;
+    }
+    return returnFireAgainstAttackingSubs;
+  }
+
+  /*
+  // Taken from MustFightBattle::returnFireAgainstDefendingSubs
+  private MustFightBattle.ReturnFire returnFireAgainstDefendingSubs(
+  final Collection<Unit> attackingUnits, final Collection<Unit> defendingUnits) {
+    final boolean attackingSubsSneakAttack =
+        defendingUnits.stream().noneMatch(Matches.unitIsDestroyer());
+    final boolean defendingSubsSneakAttack =
+    defendingSubsSneakAttackAndNoAttackingDestroyers(attackingUnits);
+    final MustFightBattle.ReturnFire returnFireAgainstDefendingSubs;
+    if (!defendingSubsSneakAttack) {
+      returnFireAgainstDefendingSubs = MustFightBattle.ReturnFire.ALL;
+    } else if (attackingSubsSneakAttack || Properties.getWW2V2(parameters.getData())) {
+      returnFireAgainstDefendingSubs = MustFightBattle.ReturnFire.SUBS;
+    } else {
+      returnFireAgainstDefendingSubs = MustFightBattle.ReturnFire.NONE;
+    }
+    return returnFireAgainstDefendingSubs;
+  }
+   */
+
+  // Taken from MustFightBattle::defendingSubsSneakAttackAndNoAttackingDestroyers
+  private boolean defendingSubsSneakAttackAndNoAttackingDestroyers(
+      final Collection<Unit> attackingUnits) {
+    return attackingUnits.stream().noneMatch(Matches.unitIsDestroyer())
+        && defendingSubsSneakAttack();
+  }
+
+  // Taken from MustFightBattle::defendingSubsSneakAttack
+  private boolean defendingSubsSneakAttack() {
+    return Properties.getWW2V2(parameters.getData())
+        || Properties.getDefendingSubsSneakAttack(parameters.getData());
+  }
+
+  @Override
+  public String toString() {
+    final String type;
+    switch (units.getType()) {
+      case AA_ATTACKER:
+        type = "AA";
+        break;
+      case AA_DEFENDER:
+        type = "AD";
+        break;
+      case SUB_ATTACKER:
+        type = "SA";
+        break;
+      case SUB_DEFENDER:
+        type = "SD";
+        break;
+      case ATTACKER:
+        type = "AT";
+        break;
+      case DEFENDER:
+        type = "DE";
+        break;
+      default:
+        type = "??";
+        break;
+    }
```

Comment:
This might be something that could get built into the Type enum.
Just give this enum a "string representation field" and call the getter here.